### ISSUE TITLE
Sigmoid and softmax exports to ONNX do not give expected results

### ISSUE THREAD

**Issue by Rayndell** [pytorch/vision#9325]
### 🐛 Describe the bug

I have troubles exporting a classification model to ONNX. My model outputs logits, which I convert into sigmoids during the evaluation, which gives fairly correct results. I was hoping to do the same in my ONNX model (find the export code below). Unfortunately the outputs are quite different from what I had during the evaluation. The sigmoid output during the evaluation looked like this:

[6.4327e-09, 6.3816e-07, 7.8797e-07, 9.9999e-01, 1.8372e-07]

And I am only able to get this kind of vectors with the ONNX model:

[1. 1. 1. 0. 0.]

Only 0s and 1s. The same happens when using the softmax function. What could possibly be wrong?

```python
class ClassificationModel(nn.Module):
    def __init__(self, base_model, transform=None):
        super(ClassificationModel, self).__init__()
        self.base_model = base_model
        self.transform = transform
        
    def forward(self, input):
        if self.transform:
            input = self.transform(input)
        output = self.base_model(input)
        output = torch.sigmoid(output)
        return output
    
    def export_to_onnx(self, path, device, input_shape):
        self.eval()
        input_names = ["input"]
        output_names = ["output"]
        dynamic_axes = {
            "input": {0: "batch_size", 2: "height", 3: "width"},
        }

        input_example = torch.randint(0, 255, (1, 3, input_shape[0], input_shape[1]), dtype=torch.uint8).to(device)
        torch.onnx.export(
            self,
            input_example,
            path,
            export_params=True,
            opset_version=16,
            do_constant_folding=True,
            input_names=input_names,
            output_names=output_names,
            dynamic_axes=dynamic_axes,
            dynamo=False
        )

def main(args):
    # Rebuild the model
    classes = CONFIG.class_names
    classes = sorted(classes)  # e.g., ['A', 'B', 'C', 'D']
    label_to_idx = {label: idx for idx, label in enumerate(classes)}
    base_model = build_model(CONFIG.model, len(classes), weights=None)
        
    # Load the given checkpoint
    checkpoint = torch.load(args.checkpoint_path, weights_only=True)
    base_model.load_state_dict(checkpoint, strict=False)
    
    # Wrap the preprocessing
    resize_shape = (CONFIG.resize_height, CONFIG.resize_width)
    crop_shape = (CONFIG.crop_height, CONFIG.crop_width)
    transform = get_transform(  resize_shape=(resize_shape if CONFIG.resize else None),
                                crop_shape=(crop_shape if CONFIG.crop else None), data_augmentation=False)
    model = ClassificationModel(base_model, transform)
    
    # Export
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    model.to(device)
    model.export_to_onnx(args.output, device=device, input_shape=(crop_shape if CONFIG.crop else resize_shape))
```

The functions build_model and get_transform are exactly the same as the ones used during the training and the evaluation, as well as the CONFIG that contains all parameters.

### Versions

Collecting environment information...
PyTorch version: 2.9.1+cu126
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Microsoft Windows 11 Professionnel (10.0.26100 64 bits)
GCC version: Could not collect
Clang version: Could not collect
CMake version: version 3.31.6
Libc version: N/A

Python version: 3.12.10 (tags/v3.12.10:0cc8128, Apr 8 2025, 12:21:36) [MSC v.1943 64 bit (AMD64)] (64-bit runtime)
Python platform: Windows-11-10.0.26100-SP0
Is CUDA available: True
CUDA runtime version: 12.6.20
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 581.80
cuDNN version: C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.6\bin\cudnn_ops64_9.dll
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

CPU:
Name: Intel(R) Core(TM) i7-8700K CPU @ 3.70GHz
Manufacturer: GenuineIntel
Family: 198
Architecture: 9
ProcessorType: 3
DeviceID: CPU0
CurrentClockSpeed: 3696
MaxClockSpeed: 3696
L2CacheSize: 1536
L2CacheSpeed: None
Revision: None

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] onnx==1.17.0
[pip3] onnx-ir==0.1.12
[pip3] onnxruntime==1.23.2
[pip3] onnxruntime-gpu==1.23.2
[pip3] onnxscript==0.5.6
[pip3] optree==0.16.0
[pip3] segmentation_models_pytorch==0.5.0
[pip3] tf2onnx==1.16.1
[pip3] torch==2.9.1+cu126
[pip3] torchaudio==2.9.1+cu126
[pip3] torchmetrics==1.8.2
[pip3] torchsummary==1.5.1
[pip3] torchvision==0.24.1+cu126
[conda] Could not collect

**Comment by Rayndell** [comment]
To be complete, I also tried to output directly the logits, and... there is definitely something wrong somewhere. While the logits look like this during the evaluation:

[ -8.4985,  -7.3544,   7.5543, -10.6103, -13.0464]

They look like this with the ONNX model:

[   561.88696  -4681.9375    3717.018   -12597.652     1133.029  ]

I took care to load the same checkpoint in both models, and the preprocessing functions used are the same.

**Comment by Rayndell** [comment]
Ok, so I figured out the problem. I saved my checkpoints like this in order to keep track of the metrics and to be able to resume a training:

```python
torch.save({
                'epoch': epoch,
                'model_state_dict': model.state_dict(),
                'optimizer_state_dict': optimizer.state_dict(),
                'val_loss': val_loss,
                'auc_history': auc_history
                }, filename)
```

So when trying to load directly the checkpoint into the state of the model like this:

```python
checkpoint = torch.load(args.checkpoint_path, weights_only=True)
base_model.load_state_dict(checkpoint, strict=False)
```

it "worked" but the weights were all messed up. The proper way to load is like this:

```python
checkpoint = torch.load(args.checkpoint_path, weights_only=True)
base_model.load_state_dict(checkpoint['model_state_dict'], strict=False)
```

Sorry for the inconvenience. However I was expecting an error message when using the first approach, is there a means to prevent this kind of "invisible" error?
